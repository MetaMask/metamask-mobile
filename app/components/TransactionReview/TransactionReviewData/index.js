import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { StyleSheet, Text, View, TextInput } from 'react-native';
import { colors, fontStyles } from '../../../styles/common';
import { strings } from '../../../../locales/i18n';

const styles = StyleSheet.create({
	overview: {
		paddingHorizontal: 16
	},
	label: {
		flex: 0,
		paddingRight: 18
	},
	labelText: {
		...fontStyles.bold,
		color: colors.gray,
		fontSize: 12
	},
	functionType: {
		...fontStyles.normal,
		color: colors.black,
		fontSize: 12,
		padding: 16
	},
	hexData: {
		...fontStyles.normal,
		backgroundColor: colors.white,
		color: colors.black,
		flex: 1,
		fontSize: 12,
		minHeight: 64,
		padding: 16
	},
	topOverviewRow: {
		borderBottomWidth: 1,
		borderColor: colors.lightGray
	},
	overviewRow: {
		paddingVertical: 15
	}
});

/**
 * Component that supports reviewing transaction data
 */
export default class TransactionReviewData extends Component {
	static propTypes = {
		/**
		 * Transaction object associated with this transaction
		 */
		transactionData: PropTypes.object,
		/**
		 * Transaction corresponding action key
		 */
		actionKey: PropTypes.string
	};

	render = () => {
		const {
			transactionData: { data },
			actionKey
		} = this.props;
		return (
			<View style={styles.overview}>
				{actionKey !== strings('transactions.tx_review_confirm') && (
					<View style={[styles.overviewRow, styles.topOverviewRow]}>
						<View style={styles.label}>
							<Text style={styles.labelText}>{strings('transaction.review_function_type')}</Text>
							<Text style={styles.functionType}>{actionKey}</Text>
						</View>
					</View>
				)}
				<View style={styles.overviewRow}>
					<View style={styles.label}>
						<Text style={styles.labelText}>{strings('transaction.review_hex_data')}:</Text>
					</View>
					<TextInput
						multiline
						placeholder={strings('transaction.optional')}
						style={styles.hexData}
						value={data}
						editable={false}
					/>
				</View>
			</View>
		);
	};
}
