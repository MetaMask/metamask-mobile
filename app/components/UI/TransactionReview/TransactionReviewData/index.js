import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { StyleSheet, Text, View, TouchableOpacity } from 'react-native';
import { KeyboardAwareScrollView } from 'react-native-keyboard-aware-scroll-view';
import IonicIcon from 'react-native-vector-icons/Ionicons';
import { fontStyles } from '../../../../styles/common';
import { strings } from '../../../../../locales/i18n';
import { connect } from 'react-redux';
import Device from '../../../../util/device';
import { ThemeContext, mockTheme } from '../../../../util/theme';
import ClipboardManager from '../../../../core/ClipboardManager';
import { showAlert } from '../../../../actions/alert';
import GlobalAlert from '../../../../components/UI/GlobalAlert';

const createStyles = (colors) =>
  StyleSheet.create({
    root: {
      paddingHorizontal: 24,
      paddingTop: 24,
      paddingBottom: Device.isIphoneX() ? 48 : 24,
    },
    dataHeader: {
      flexDirection: 'row',
      justifyContent: 'space-between',
      alignItems: 'center',
      width: '100%',
      marginBottom: 28,
    },
    dataTitleText: {
      ...fontStyles.bold,
      color: colors.text.default,
      fontSize: 14,
      alignSelf: 'center',
    },
    dataDescription: {
      textAlign: 'center',
      ...fontStyles.normal,
      color: colors.text.default,
      fontSize: 14,
      marginBottom: 28,
    },
    dataBox: {
      padding: 12,
      borderWidth: 1,
      borderColor: colors.border.default,
      borderRadius: 8,
      flex: 1,
    },
    label: {
      flexDirection: 'row',
      justifyContent: 'flex-start',
      marginBottom: 12,
    },
    boldLabel: {
      ...fontStyles.bold,
    },
    labelText: {
      ...fontStyles.normal,
      color: colors.text.default,
      fontSize: 14,
    },
    hexData: {
      ...fontStyles.normal,
      backgroundColor: colors.background.default,
      color: colors.text.default,
      fontSize: 14,
      paddingTop: 0,
    },
    scrollView: {
      flex: 1,
    },
  });

/**
 * PureComponent that supports reviewing transaction data
 */
class TransactionReviewData extends PureComponent {
  static propTypes = {
    /**
     * Transaction object associated with this transaction
     */
    transaction: PropTypes.object,
    /**
     * Transaction corresponding action key
     */
    actionKey: PropTypes.string,
    /**
     * Hides or shows transaction data
     */
    toggleDataView: PropTypes.func,
    /**
     * Height of custom gas and data modal
     */
    customGasHeight: PropTypes.number,
    /**
     * Triggers global alert
     */
    showAlert: PropTypes.func,
  };

  applyRootHeight = () => ({ height: this.props.customGasHeight });

  handleCopyHex = () => {
    const {
      transaction: {
        transaction: { data },
      },
    } = this.props;
    ClipboardManager.setString(data);
    this.props.showAlert({
      isVisible: true,
      autodismiss: 1500,
      content: 'clipboard-alert',
      data: { msg: strings('transaction.hex_data_copied') },
    });
  };

  render = () => {
    const {
      transaction: {
        transaction: { data },
      },
      actionKey,
      toggleDataView,
    } = this.props;
    const colors = this.context.colors || mockTheme.colors;
    const styles = createStyles(colors);

    return (
      <View style={[styles.root, this.applyRootHeight()]}>
        <View style={styles.dataHeader}>
          <TouchableOpacity onPress={toggleDataView}>
            <IonicIcon
              name={'ios-arrow-back'}
              size={24}
              color={colors.text.default}
            />
          </TouchableOpacity>
          <Text style={styles.dataTitleText}>
            {strings('transaction.data')}
          </Text>
          <IonicIcon
            name={'ios-arrow-back'}
            size={24}
            color={colors.background.default}
          />
        </View>
        <Text style={styles.dataDescription}>
          {strings('transaction.data_description')}
        </Text>
        <View style={styles.dataBox}>
          {actionKey !== strings('transactions.tx_review_confirm') && (
            <View style={styles.label}>
              <Text style={[styles.labelText, styles.boldLabel]}>
                {strings('transaction.review_function')}:{' '}
              </Text>
              <Text style={styles.labelText}>{actionKey}</Text>
            </View>
          )}
          <Text style={[styles.labelText, styles.boldLabel]}>
            {strings('transaction.review_hex_data')}:{' '}
          </Text>
          <View style={styles.scrollView}>
            <KeyboardAwareScrollView style={styles.scrollView}>
              <TouchableOpacity
                activeOpacity={0.8}
                onPress={this.handleCopyHex}
                style={styles.scrollView}
              >
                <Text style={styles.hexData}>{data}</Text>
              </TouchableOpacity>
            </KeyboardAwareScrollView>
          </View>
        </View>
        <GlobalAlert />
      </View>
    );
  };
}

const mapStateToProps = (state) => ({
  conversionRate:
    state.engine.backgroundState.CurrencyRateController.conversionRate,
  currentCurrency:
    state.engine.backgroundState.CurrencyRateController.currentCurrency,
  contractExchangeRates:
    state.engine.backgroundState.TokenRatesController.contractExchangeRates,
  transaction: state.transaction,
});

const mapDispatchToProps = (dispatch) => ({
  showAlert: (config) => dispatch(showAlert(config)),
});

TransactionReviewData.contextType = ThemeContext;

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(TransactionReviewData);
