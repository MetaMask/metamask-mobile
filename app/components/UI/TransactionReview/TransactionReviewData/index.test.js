import React from 'react';
import TransactionReviewData from './';
import configureMockStore from 'redux-mock-store';
import { shallow } from 'enzyme';

const mockStore = configureMockStore();

describe('TransactionReviewData', () => {
	it('should render correctly', () => {
		const initialState = {
			engine: {
				backgroundState: {
					TokenRatesController: {
						contractExchangeRates: {}
					},
					CurrencyRateController: {
						currentCurrency: 'usd',
						conversionRate: 0.1
					}
				}
			},
			transaction: {
				transaction: {
					data: ''
				},
				value: '',
				from: '0x1',
				gas: '',
				gasPrice: '',
				to: '0x2',
				selectedAsset: undefined,
				assetType: undefined
			}
		};

		const wrapper = shallow(
			<TransactionReviewData transactionData={{ amount: 0, gas: 0, gasPrice: 1, from: '0x0' }} />,
			{
				context: { store: mockStore(initialState) }
			}
		);
		expect(wrapper.dive()).toMatchSnapshot();
	});
});
