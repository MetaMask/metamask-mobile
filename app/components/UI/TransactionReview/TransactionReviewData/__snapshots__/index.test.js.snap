// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionReviewData should render correctly 1`] = `
<View
  style={
    Array [
      Object {
        "paddingBottom": 48,
        "paddingHorizontal": 24,
        "paddingTop": 24,
      },
      Object {
        "height": undefined,
      },
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 28,
        "width": "100%",
      }
    }
  >
    <TouchableOpacity>
      <Icon
        allowFontScaling={false}
        color="#24292E"
        name="ios-arrow-back"
        size={24}
      />
    </TouchableOpacity>
    <Text
      style={
        Object {
          "alignSelf": "center",
          "color": "#24292E",
          "fontFamily": "EuclidCircularB-Bold",
          "fontSize": 14,
          "fontWeight": "600",
        }
      }
    >
      Data
    </Text>
    <Icon
      allowFontScaling={false}
      color="#FFFFFF"
      name="ios-arrow-back"
      size={24}
    />
  </View>
  <Text
    style={
      Object {
        "color": "#24292E",
        "fontFamily": "EuclidCircularB-Regular",
        "fontSize": 14,
        "fontWeight": "400",
        "marginBottom": 28,
        "textAlign": "center",
      }
    }
  >
    Data associated with this transaction
  </Text>
  <View
    style={
      Object {
        "borderColor": "#bbc0c5",
        "borderRadius": 8,
        "borderWidth": 1,
        "flex": 1,
        "padding": 12,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "flex-start",
          "marginBottom": 12,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#24292E",
              "fontFamily": "EuclidCircularB-Regular",
              "fontSize": 14,
              "fontWeight": "400",
            },
            Object {
              "fontFamily": "EuclidCircularB-Bold",
              "fontWeight": "600",
            },
          ]
        }
      >
        Function
        :
         
      </Text>
      <Text
        style={
          Object {
            "color": "#24292E",
            "fontFamily": "EuclidCircularB-Regular",
            "fontSize": 14,
            "fontWeight": "400",
          }
        }
      />
    </View>
    <Text
      style={
        Array [
          Object {
            "color": "#24292E",
            "fontFamily": "EuclidCircularB-Regular",
            "fontSize": 14,
            "fontWeight": "400",
          },
          Object {
            "fontFamily": "EuclidCircularB-Bold",
            "fontWeight": "600",
          },
        ]
      }
    >
      Hex data
      : 
    </Text>
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <KeyboardAwareScrollView
        enableAutomaticScroll={true}
        enableOnAndroid={false}
        enableResetScrollToCoords={true}
        extraHeight={75}
        extraScrollHeight={0}
        keyboardOpeningTime={250}
        style={
          Object {
            "flex": 1,
          }
        }
        viewIsInsideTabBar={false}
      >
        <TouchableWithoutFeedback
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "backgroundColor": "#FFFFFF",
                "color": "#24292E",
                "fontFamily": "EuclidCircularB-Regular",
                "fontSize": 14,
                "fontWeight": "400",
                "paddingTop": 0,
              }
            }
          />
        </TouchableWithoutFeedback>
      </KeyboardAwareScrollView>
    </View>
  </View>
</View>
`;
